design/bytecode.md

```diff
@@ -0,0 +1,106 @@
+# Byte code
+
+The Byte code of Seed is generated from the SeedIR code, and executed by the
+SeedInterpreter as a virtual machine. The byte code instructions are 32 bits in
+length. All instructions have an opcode in the first 7 bits, so there are at
+most 128 kinds of opcode. Other fields of the instructions are listed below:
+
+- A (8 bits): The operand A
+- B (8 bits): The operand B
+- C (8 bits): The operand C
+- K (1 bit): Indicates if the operand C is the index of the register (K = 0) or
+  constant (K = 1)
+- Bx (17 bits): The operand Bx, combined with B, C and K
+- sBx (17 bits): The operand sBx, A signed argument in the range of [-Max, Max)
+  (Max = 2 ^ 16). The actual value is the unsigned value (Bx) minus Max, so that
+  -Max is represented by 0, and Max - 1 is represented by 2 \* Max - 1.
+
+The layout of the instructions is illustrated as below:
+
+```text
+
+     3 3 2 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0
+     1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+    |K|      C (8)    |      B (8)    |     A (8)     |   Op (7)    |
+    |             Bx (17)             |     A (8)     |   Op (7)    |
+    |            sBx (17)             |     A (8)     |   Op (7)    |
+
+```
+
+SeedInterpreter is a register based virtual machine. There are maximum 255
+registers for each function. The local variables of a function is pre-allocated
+in the registers by the compiler. So the user could not define more than 255
+local variables for a certain function.
+
+Each function has its own constant list, which is generated by the compiler. The
+elements in the constant list are usually numbers and strings without
+duplication.
+
+## Instruction Notation
+
+Following notations are used to explain the behavior of the instructions.
+
+- PC: Program Counter
+- R[A]: Register at index A
+- R[B]: Register at index B
+- R[C]: Register at index C
+- K[N]: Element N in the constant list
+- RK[C]: Register at index C (when K = 0) or constant at index C (K = 1)
+
+## Move Instruction
+
+```text
+Move A B            // R[A] := R[B]
+```
+
+Copy the value of the register R[B] into register R[A]. If R[B] holds a object,
+the reference to that object is copied. Move is often used for moving values
+into place for the next operation.
+
+## LoadI Instruction
+
+```text
+LoadI A sBx         // R[A] := sBx
```

Comment:
Is this an equivalent but a quicker (immediate value) version of LoadK?

Reply:
It's just a optimization for LoadK to reduce the size of constant list. And the speed will be a little quicker.